specifications/Disruptor/Disruptor_SPMC.tla

```diff
@@ -0,0 +1,155 @@
+----------------------------- MODULE Disruptor_SPMC -------------------------
+(***************************************************************************)
+(* Models a Single Producer, Multi Consumer Disruptor (SPMC).              *)
+(*                                                                         *)
+(* The model verifies that no data races occur between the publisher       *)
+(* and consumers and that all consumers eventually read all published      *)
+(* values.                                                                 *)
+(*                                                                         *)
+(* To see a data race, try and run the model with two publishers.          *)
+(***************************************************************************)
+
+EXTENDS Naturals, Integers, FiniteSets, Sequences
+
+CONSTANTS
+  Writers,      (* Writer/publisher thread ids.    *)
+  Readers,      (* Reader/consumer  thread ids.    *)
+  MaxPublished, (* Max number of published events. *)
```

Comment:
I presume this constant is only relevant for model checking? It would be cleaner to separate the logical spec from the bounds imposed for model checking and either add a state constraint such as ```published < MaxPublished``` in the cfg file or write a MC version of the spec that adds extra guards to actions.

Reply:
Yes and no. I've investigated and the constant does bound the model. So I agree - that would be better to have as a state constraint. But then I can't model check the liveliness property in the model (that all consumers eventually always read all published events) as liveliness cannot be verified when state or action constraints are specified.
Is there a "clean" third option that you know of?

Reply:
Apologies for the late reply: now I am confused. First, I would rather have defined the liveness property as
```
Liveliness == \A r \in Readers : \A i \in 1 .. MaxPublished :
  <>[](i \in 1 .. published => Len(consumed[r]) >= i /\ consumed[r][i] = i-1)
```
(with the second conjunct on the right-hand side being optional). The reason is that the property that you assert will not hold if writers are allowed to continue publishing even after reaching the (artificial) bound since readers would then be able to update their ```consumed``` sequence as well, whereas the above property should always hold. (Even more general, the bound on ```i``` could be ```Nat``` but that would require an override of ```Nat``` during model checking so that TLC doesn't complain about an infinite quantifier bound, and that would be a little heavy.)

I then tried commenting out the guard ```next < MaxPublished``` in ```BeginWrite``` and adding the constraint
```
StateConstraint == published <= MaxPublished
```
for model checking. This appears to work just fine, and when I comment out the fairness condition on ```BeginRead```, TLC gives me the expected failure of the temporal property.

I would even suggest removing the fairness condition on ```BeginWrite``` from the spec, since why should writers be required to publish items forever? The above liveness property (but of course not the original one) continues to hold for the modified spec.

Of course, it's up to you to decide what you intend as your specification.

Reply:
Thanks for the detailed description. I'll dive deeper into it in the weekend when I have more time. I'm keen on improving the spec (and learn!) :-)

Reply:
Stephan's `Liveliness` property would allow to model `consumed` as a counter instead of a sequence (for each reader). This observation also led me to realize that the SPMC specification seems to describe the [multicast configuration](https://lmax-exchange.github.io/disruptor/disruptor.htm), since all readers are consuming every value. It might be helpful to include this in a comment for clarity.

Reply:
If `consumed` remains a sequence, you could add the following "action property" (with `IsPrefix` defined at https://github.com/tlaplus/CommunityModules/blob/master/modules/SequencesExt.tla#L219-L225 :

```tla
\* We only ever append to the history variable  consumed  .   
Increments ==
    [][\A r \in Readers: IsPrefix(consumed[r], consumed'[r])]_vars
```

Reply:
> Stephan's `Liveliness` property would allow to model `consumed` as a counter instead of a sequence (for each reader). This observation also led me to realize that the SPMC specification seems to describe the [multicast configuration](https://lmax-exchange.github.io/disruptor/disruptor.htm), since all readers are consuming every value. It might be helpful to include this in a comment for clarity.

Hi @lemmy, I wrote in the top that it's a Single Producer Multiple Consumer Disruptor - that's where I thought I made it clear that it is *that* configuration of the Disruptor. But perhaps I am assuming that everyone knows all consumers read all events (i.e. "multicast" behaviour) and that is not common knowledge?

Reply:
> Apologies for the late reply: now I am confused. First, I would rather have defined the liveness property as
> 
> ```
> Liveliness == \A r \in Readers : \A i \in 1 .. MaxPublished :
>   <>[](i \in 1 .. published => Len(consumed[r]) >= i /\ consumed[r][i] = i-1)
> ```
> 
> (with the second conjunct on the right-hand side being optional). The reason is that the property that you assert will not hold if writers are allowed to continue publishing even after reaching the (artificial) bound since readers would then be able to update their `consumed` sequence as well, whereas the above property should always hold. (Even more general, the bound on `i` could be `Nat` but that would require an override of `Nat` during model checking so that TLC doesn't complain about an infinite quantifier bound, and that would be a little heavy.)
> 
> I then tried commenting out the guard `next < MaxPublished` in `BeginWrite` and adding the constraint
> 
> ```
> StateConstraint == published <= MaxPublished
> ```
> 
> for model checking. This appears to work just fine, and when I comment out the fairness condition on `BeginRead`, TLC gives me the expected failure of the temporal property.
> 
> I would even suggest removing the fairness condition on `BeginWrite` from the spec, since why should writers be required to publish items forever? The above liveness property (but of course not the original one) continues to hold for the modified spec.
> 
> Of course, it's up to you to decide what you intend as your specification.

I did the changes (see a new PR) and I quite like it: Getting rid of the artificial model constraint in the `BeginWrite` action, remove the requirement to have producers publish forever and moving the bounding of the model to a state constraint. Elegant and a better model.

I still have some work to do for the MPMC model as it's more complex to write the liveliness property because there's multiple producers and hence it more difficult to express what `sequence` number a consumer can actually read. (More details will follow).

Reply:
Glad to hear that it worked out! I didn't look very much at the MPMC model, please let me know if you want me to.